language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/scopeproviders/ServiceScopeProvider.java

```diff
@@ -0,0 +1,105 @@
+
+import org.antlr.v4.runtime.CommonToken;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.ballerinalang.annotation.JavaSPIService;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.compiler.DocumentServiceKeys;
+import org.ballerinalang.langserver.compiler.LSContext;
+import org.ballerinalang.langserver.completions.CompletionKeys;
+import org.ballerinalang.langserver.completions.CompletionSubRuleParser;
+import org.ballerinalang.langserver.completions.providers.contextproviders.AnnotationAttachmentContextProvider;
+import org.ballerinalang.langserver.completions.spi.LSCompletionProvider;
+import org.ballerinalang.langserver.completions.util.Snippet;
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4j.Position;
+import org.wso2.ballerinalang.compiler.parser.antlr4.BallerinaParser;
+import org.wso2.ballerinalang.compiler.tree.BLangNode;
+import org.wso2.ballerinalang.compiler.tree.BLangService;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangExpression;
+import org.wso2.ballerinalang.compiler.util.diagnotic.DiagnosticPos;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * ServiceContextResolver.
+ */
+@JavaSPIService("org.ballerinalang.langserver.completions.spi.LSCompletionProvider")
+public class ServiceScopeProvider extends LSCompletionProvider {
+
+    public ServiceScopeProvider() {
+        this.attachmentPoints.add(BLangService.class);
+    }
+
+    @Override
+    public List<CompletionItem> getCompletions(LSContext ctx) {
+        ArrayList<CompletionItem> completionItems = new ArrayList<>();
+        if (this.isWithinAttachedExpressions(ctx)) {
+            // suggest all the visible, defined listeners
+            return this.getCompletionItemsAfterOnKeyword(ctx);
+        }
+        List<CommonToken> lhsTokens = ctx.get(CompletionKeys.LHS_TOKENS_KEY);
+        if (lhsTokens != null && !lhsTokens.isEmpty()
+                && lhsTokens.get(lhsTokens.size() - 1).getType() == BallerinaParser.AT) {
+            return this.getProvider(AnnotationAttachmentContextProvider.class).getCompletions(ctx);
+        }
+
+        Optional<String> subRule = this.getSubrule(lhsTokens);
+        subRule.ifPresent(rule -> CompletionSubRuleParser.parseWithinServiceDefinition(rule, ctx));
+        ParserRuleContext parserRuleContext = ctx.get(CompletionKeys.PARSER_RULE_CONTEXT_KEY);
+
+        if (parserRuleContext != null && this.getProvider(parserRuleContext.getClass()) != null) {
+            return this.getProvider(parserRuleContext.getClass()).getCompletions(ctx);
+        }
+
+        completionItems.addAll(this.getBasicTypes(ctx.get(CompletionKeys.VISIBLE_SYMBOLS_KEY)));
+        completionItems.addAll(this.getPackagesCompletionItems(ctx));
+        completionItems.add(Snippet.KW_PUBLIC.get().build(ctx));
+        Optional<List<CompletionItem>> resourceSnippets = this.getResourceSnippets(ctx);
```

Comment:
if `getResourceSnippets(ctx)` returns a `List<CompletionItem>`; we can return an empty list when there's no resource snippets. Then we don't need `.ifPresent()` check for the list since `.addAll()` internally does it. WDYT?

```java
List<CompletionItem> resourceSnippets = this.getResourceSnippets(ctx);
completionItems.addAll(resourceSnippets);
```

Reply:
resolved